File: 2024/day12/day12.hs
import System.IO
import Data.Set as Data
import Data.List

one_step:: (Int, Int)->Int->Int->[(Int, Int)]
one_step (x, y) rows cols = [n|n<-[(x-1, y), (x, y+1), (x+1, y), (x, y-1)], 0<= fst n, fst n<rows, 0<= snd n, snd n<cols]

fixed_point :: (Eq a)=>(a->a)->a->a
fixed_point f x =
    if ((f x) == x) then x
    else fixed_point f (f x)

update :: [[Char]]->Int->Int->[(Int, Int)]->[(Int, Int)]
update grid rows cols region =
    if region == [] then []
    else
        let
            label = (grid!!(fst rnd)!!(snd rnd)) where rnd = head region
            to_add = concat [[n|n<-(one_step x rows cols), ((grid!!fst n)!!snd n) == label]|x<-region]
        in
            nub (region++to_add)

cost :: [(Int, Int)]->Int
cost region=
    let
        area = length region
        perimeter = sum [length [n|n<-[(x-1, y), (x, y+1), (x+1, y), (x, y-1)], not (n `elem` region)]|(x, y)<-region]
    in
        area*perimeter

is_corner :: [(Int, Int)]->(Int, Int)->Int->Int->(Bool, Int)
is_corner region place rows cols =
    let
        neighbours = [(x, y)|(x, y)<-(one_step place rows cols), ((x, y)`elem` region)]
        x = fst place
        y = snd place
    in
        if (length neighbours == 4)||(length neighbours == 3) then (False, 0)
        else if (length neighbours == 2) then
            let
                n1 = neighbours!!0
                n2 = neighbours!!1
            in
                if (fst n1 == fst n2)||(snd n1 == snd n2) then (False, 0)
                else let
                    ver = head [n|n<-neighbours, snd n == y]
                    hor = head [n|n<-neighbours, fst n == x]
                    diag = if ((x- fst ver)*(y-snd hor) == 1) then [n|n<-[(x-1, y+1), (x+1, y-1)], n`elem`region, 0<= fst n, fst n<rows, 0<= snd n, snd n<cols] else [n|n<-[(x+1, y+1), (x-1, y-1)], n`elem`region, 0<= fst n, fst n<rows, 0<= snd n, snd n<cols]
                in
                    if (length diag == 0) then (True, -2)
                    else if (length diag == 1) then (True, 0)
                    else (True, 2)
        else if (length neighbours == 1) then
            let
                parent = head neighbours
                corners = if (fst parent == x) then [n|n<-[(fst parent-1, snd parent), (fst parent+1, snd parent)], n`elem`region, 0<= fst n, fst n<rows, 0<= snd n, snd n<cols] else [n|n<-[(fst parent, snd parent-1), (fst parent, snd parent+1)], n`elem`region, 0<= fst n, fst n<rows, 0<= snd n, snd n<cols]
            in
                if (length corners == 0) then (True, 0)
                else if (length corners == 1) then (True, 2)
                else (True, 4)
        else (True, 4)

sides :: [(Int, Int)] -> Int -> Int -> Int
sides region rows cols =
    if region == [] then 0
    else
        if length region == 1 then 4
        else
            let
                to_remove = head (dropWhile (\x-> not(fst (is_corner region x rows cols))) region)
                change = snd (is_corner region to_remove rows cols)
            in
                change+(sides (Data.List.filter (/=to_remove) region) rows cols)


all_regions :: [[Char]] -> [(Int, Int)] -> [(Int, Int)] -> [[(Int, Int)]] -> Int -> Int -> [[(Int, Int)]]
all_regions grid positions_seen positions_unseen regions rows cols =
    if (length positions_unseen == 0) then regions
    else
        let
            new_region = fixed_point (update grid rows cols) [head positions_unseen]
        in
            all_regions grid (positions_seen++new_region) [x|x<-positions_unseen, not (x `elem` new_region)] (regions++[new_region]) rows cols

main = do
    handle <- openFile "day12.txt" ReadMode
    contents <- hGetContents handle
    let 
        grid = lines contents
        rnum = length grid
        cnum = length (grid!!0)
        a = all_regions grid [] [(x, y)|x<-[0..rnum-1], y<-[0..cnum-1]] [] rnum cnum
    print(length a)
    print(sum [cost x|x<-a])
    print(sum [length x*(sides x rnum cnum)|x<-a])

File: 2024/day12/day12.py
def update(grid, rows, cols, region):
    if region == []:
        return []
    else:
        to_add = set([])
        random = region.pop()
        region.add(random)
        label = grid[random[0]][random[1]]
        for place in region:
            x = place[0]
            y = place[1]
            for n in [(x-1, y), (x, y+1), (x+1, y), (x, y-1)]:
                a = n[0]
                b = n[1]
                if (0<=a<rows) and (0<=b<cols) and (grid[a][b] == label):
                    to_add.add(n)
        return region.union(to_add)

def cost(region):
    area = len(region)
    perimeter = 0
    for place in region:
        x = place[0]
        y = place[1]
        for n in [(x-1, y), (x, y+1), (x+1, y), (x, y-1)]:
            if n not in region:
                perimeter += 1
    return area*perimeter

def sides(region, rows, cols):
    if region == set([]):
        return 0
    elif len(region) == 1:
        return 4
    else:
        to_remove = (0, 0)
        change = 0
        for place in region:
            x = place[0]
            y = place[1]
            one_step = [n for n in [(x-1, y), (x, y+1), (x+1, y), (x, y-1)] if n in region and (0<=n[0]<rows) and (0<=n[1]<cols)]
            if len(one_step) == 4:
                continue
            elif len(one_step) == 3:
                continue
            elif len(one_step) == 2:
                if (one_step[0][0]==one_step[1][0]) or (one_step[0][1]==one_step[1][1]):
                    continue
                else:
                    ver = [n for n in one_step if n[1]==y][0]
                    hor = [n for n in one_step if n[0]==x][0]
                    if (x-ver[0])*(y-hor[1]) == 1:
                        diag = [n for n in [(x-1, y+1), (x+1, y-1)] if n in region and (0<=n[0]<rows) and (0<=n[1]<cols)]
                    else:
                        diag = [n for n in [(x-1, y-1), (x+1, y+1)] if n in region and (0<=n[0]<rows) and (0<=n[1]<cols)]
                    if len(diag) == 0:
                        to_remove = place
                        change = -2
                        break
                    elif len(diag) == 1:
                        to_remove = place
                        change = 0
                        break
                    else:
                        to_remove = place
                        change = 2
                        break
            elif len(one_step) == 1:
                parent = one_step[0]
                if parent[0]==x:
                    corners = [n for n in [(parent[0]-1, parent[1]), (parent[0]+1, parent[1])]if n in region and (0<=n[0]<rows) and (0<=n[1]<cols)]
                else:
                    corners = [n for n in [(parent[0], parent[1]-1), (parent[0], parent[1]+1)]if n in region and (0<=n[0]<rows) and (0<=n[1]<cols)]
                if len(corners) == 0:
                    to_remove = place
                    change = 0
                    break
                elif len(corners) == 1:
                    to_remove = place
                    change = 2
                    break
                else:
                    to_remove = place
                    change = 4
                    break
            else:
                to_remove = place
                change = 4
                break
        region.remove(to_remove)
        return change+sides(region, rows, cols)
        
def total_cost(grid, rows, cols):
    all_pos = [(x, y) for x in range(rows) for y in range(cols)]
    all_regions = []
    while all_pos:
        seed = set([all_pos[0]])
        updated = update(grid, rows, cols, seed)
        while seed != updated:
            seed = updated
            updated = update(grid, rows, cols, seed)
        all_regions.append(seed)
        for item in seed:
            all_pos.remove(item)
    return sum([cost(x) for x in all_regions]), sum([len(x)*sides(x, rows, cols) for x in all_regions])

with open("day12.txt", "r") as f:
    grid = f.read().split("\n")
rnum = len(grid)
cnum = len(grid[0])
print(total_cost(grid, rnum, cnum))